identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccess.kt

```diff
@@ -0,0 +1,493 @@
+    private const val CMD_MDOC_DELETE_CREDENTIAL = 0x08
+    private const val CMD_MDOC_PROVISION_DATA = 0x09
+    private const val CMD_MDOC_SWAP_IN = 0x06
+    private const val APDU_RESPONSE_STATUS_OK = 0x9000
+    private const val INS_ENVELOPE = 0xC3.toByte()
+
+    val transport = DirectAccessOmapiTransport()
+
+    /**
+     * Returns the maximum size, in bytes, of `credentialData` which can be used in credentials.
+     * This will be at least 32 KiB.
+     */
+    val maximumCredentialSize: Long
+        get() {
+            // todo integrate with applet
+            return 32768
+        }
+
+    /**
+     * Returns true if direct access is supported on the device, false otherwise.
+     */
+    val isDirectAccessSupported: Boolean
+        get() {
+            try {
+                transport.openConnection()
+            } catch (e: Throwable) {
+                Logger.d(TAG, "${e.message}")
+                return false
+            }
+            transport.closeConnection()
+            return true
+        }
+
+    private fun setShort(buf: ByteArray, offset: Int, value: Short) {
+        buf[offset] = (value.toInt() shr 8 and 0xFF).toByte()
+        buf[offset + 1] = (value.toInt() and 0xFF).toByte()
+    }
+
+    private fun getAPDUResponseStatus(input: ByteArray): Int {
+        // Last two bytes are the status SW0SW1
+        val SW0 = input[input.size - 2]
+        val SW1 = input[input.size - 1]
+        return SW0.toInt() shl 8 or SW1.toInt() and 0xFFFF
+    }
+
+    @Throws(IOException::class)
+    private fun makeCommandApdu(data: ByteArray): ByteArray {
+        // TODO Handle non extended length.
+        val bos = ByteArrayOutputStream()
+        bos.write(0) // CLS
+        bos.write(INS_ENVELOPE.toInt()) // INS
+        bos.write(0) // P1
+        bos.write(0) // P2
+        // Send extended length APDU always as response size is not known to HAL.
+        // Case 1: Lc > 0  CLS | INS | P1 | P2 | 00 | 2 bytes of Lc | CommandData | 2 bytes of Le
+        // all set to 00.
+        // Case 2: Lc = 0  CLS | INS | P1 | P2 | 3 bytes of Le all set to 00.
+        bos.write(0x00)
+        // Extended length 3 bytes, starts with 0x00
+        if (data.isNotEmpty()) {
+            bos.write(data.size shr 8)
+            bos.write(data.size and 0xFF)
+            // Data
+            bos.write(data)
+        }
+        bos.write(0)
+        bos.write(0)
+        return bos.toByteArray()
+    }
+
+    @Throws(IOException::class)
+    private fun sendApdu(
+        cmd: Int,
+        slot: Int,
+        data: ByteArray,
+        offset: Int,
+        length: Int,
+        operation: Byte,
+    ): ByteArray? {
+        val bb = ByteBuffer.allocate(length)
+        bb.put(data, offset, length)
+        val scratchpad = ByteArray(2)
+        val bos = ByteArrayOutputStream()
+        // set instruction
+        setShort(scratchpad, 0, cmd.toShort())
+        bos.write(scratchpad)
+        bos.write(slot)
```

Comment:
Check that slot >=0 and <= 0xff, so this doesn't write something unexpected.
...or should that be 0..127, rather than 0..255?

Reply:
Agreed. Applet expects the slot as byte value. We will add this check.